NitroxServer/GameLogic/PlayerManager.cs

```diff
@@ -17,20 +21,58 @@ public List<Player> GetPlayers()
             }
         }
 
-        public Player PlayerAuthenticated(Connection connection, string playerId)
+        public PlayerSlotReservation ReservePlayerSlot(string correlationId, string playerId)
```

Comment:
You might want to consider adding some documentation to the reservation system. It took me a little while to figure out that the client does a reservation, then receives whether that reservation succeeded, and then has to respond to that again before actually ending up in the session.

The reason for doing so isn't entirely clear. Is this to allow players to correct their name (and more stuff in the future) while still having thar reserved spot no-one else can occupy (for a limited amount of time)?

Reply:
The process of reserving a slot gives us a lot of options for how players join servers. The workflow allows us to tell a user if they'll be able to join the session before going through all of the steps to try and do so. With the old workflow, the game would have fully started before we would simply kick them out - yucky UX.

Now, players send information to the server in order to see if they can get a spot. Currently, the only prerequisite is that they join with a unique user name. In the future, we may do things like also check the player limit to see if there is enough room or check an IP ban list to see if this person is a troublemaker that has been exiled before letting them become a "player" on the server. 

The reason that we do the reservation on success is because the user does not actually join the session until the game has fully loaded - which takes a little while. That is ubiquitous with the old workflow. The reservation makes sure that we do not end up in a race condition between a user that is loading up their game when another comes along and tries to join the same server with the same name (or meets any other criteria that would result in a rejection.)

This also allows us to handle persistent reservations for games hosts and other privileged players if we go down that route.

I think I called out the potential for "YAGNI" elsewhere, so forgive me if this all seems a little too pie-in-the-sky for the original vision of this project. I just saw an opportunity to roll all of those common concerns into a consistent pipeline and jumped on it.

Reply:
Hehe, I totally forgot about the race condition between when the player initiates the join, and has the world loaded and is ready to spawn and play (and send/receive game updates). The reservation makes sense now, thanks!